src/sentry/workflow_engine/processors/detector.py

```diff
@@ -55,6 +57,97 @@ def get_detector_by_event(event_data: WorkflowEventData) -> Detector:
     return detector
 
 
+def get_detectors_by_events_bulk(
+    event_list: list[GroupEvent],
+) -> Mapping[str, Detector]:
+    if not event_list:
+        return {}
+
+    # Separate events by whether they have occurrences or not
+    events_with_occurrences: list[tuple[GroupEvent, IssueOccurrence]] = []
+    events_without_occurrences: list[GroupEvent] = []
+
+    for event in event_list:
+        issue_occurrence = event.occurrence
+
+        if issue_occurrence is None:
+            events_without_occurrences.append(event)
+        else:
+            events_with_occurrences.append((event, issue_occurrence))
+
+    result: dict[str, Detector] = {}
+
+    # Fetch detectors for events with occurrences (by detector_id)
+    if events_with_occurrences:
+        detector_ids = []
+        event_to_detector_id: dict[GroupEvent, int] = {}
+
+        for event, occurrence in events_with_occurrences:
+            detector_id = occurrence.evidence_data.get("detector_id")
+            if detector_id is not None:
+                detector_ids.append(detector_id)
+                event_to_detector_id[event] = detector_id
+
+        if detector_ids:
+            detectors_by_id = {
+                detector.id: detector for detector in Detector.objects.filter(id__in=detector_ids)
+            }
+
+            for event, detector_id in event_to_detector_id.items():
+                detector = detectors_by_id.get(detector_id)
+                if detector:
+                    result[event.event_id] = detector
+                else:
+                    metrics.incr("workflow_engine.detectors.error")
+                    logger.warning(
+                        "Detector not found for event",
+                        extra={
+                            "event_id": event.event_id,
+                            "group_id": event.group_id,
+                            "detector_id": detector_id,
+                        },
+                    )
+
+    # Fetch detectors for events without occurrences (by project_id and issue_type)
+    if events_without_occurrences:
```

Comment:
🤔 is the only time we'd expect this for the `ErrorDetector`? should we add more checks to ensure that we only handle this case for the error detector? (nervous about over executing the detectors)

Reply:
If the error detector is the only case where there isn't an occurrence, then yes. I'm pretty sure this is the case...